src/systems/filecoin_markets/storage_market/storage_market_actor_code.go

```diff
@@ -0,0 +1,278 @@
+		}
+	}
+
+	UpdateRelease(rt, h, st)
+
+	return response
+}
+
+func (a *StorageMarketActorCode_I) VerifyPublishedDealIDs(rt Runtime, dealIDs []deal.DealID) {
+
+	TODO() // verify StorageMinerActor
+
+	h, st := a.State(rt)
+
+	for _, dealID := range dealIDs {
+
+		publishedDeal := st._getOnchainDeal(rt, dealID)
+		st._assertPublishedDealState(rt, dealID)
+
+		dealP := publishedDeal.Deal().Proposal()
+		st._assertDealStartAfterCurrEpoch(rt, dealP)
+
+		// deal must not expire before the maximum allowable epoch between pre and prove commits
+		// we do not have to check if the deal has expired at ProveCommit
+		// if the MAX_PROVE_COMMIT_SECTOR_EPOCH constraint is not violated
+		st._assertDealExpireAfterMaxProveCommitWindow(rt, dealP)
+
+	}
+
+	Release(rt, h, st)
+}
+
+func (a *StorageMarketActorCode_I) ActivateDeals(rt Runtime, dealIDs []deal.DealID) []deal.OnchainDeal {
+
+	TODO() // verify StorageMinerActor
+
+	h, st := a.State(rt)
+	ret := make([]deal.OnchainDeal, len(dealIDs))
+
+	for _, dealID := range dealIDs {
+		publishedDeal := st._getOnchainDeal(rt, dealID)
+		st._assertPublishedDealState(rt, dealID)
+
+		dealP := publishedDeal.Deal().Proposal()
+		st._assertDealNotYetExpired(rt, dealP)
+
+		onchainDeal := st._activateDeal(rt, publishedDeal)
+		ret = append(ret, onchainDeal)
+	}
+
+	UpdateRelease(rt, h, st)
+
+	return ret
+
+}
+
+func (a *StorageMarketActorCode_I) ProcessDealSlash(rt Runtime, dealIDs []deal.DealID, slashAction deal.StorageDealSlashAction) {
+
+	TODO() // only call by StorageMinerActor
+
+	h, st := a.State(rt)
+
+	// only terminated fault will result in slashing of deal collateral
+
+	switch slashAction {
+	case deal.SlashTerminatedFaults:
+		st._slashTerminatedFaults(rt, dealIDs)
+	default:
+		rt.Abort("sma.ProcessDealSlash: invalid action type")
+	}
+
+	UpdateRelease(rt, h, st)
+
+}
+
+func (a *StorageMarketActorCode_I) ProcessDealPayment(rt Runtime, dealIDs []deal.DealID, newPaymentEpoch block.ChainEpoch) {
+	h, st := a.State(rt)
+
+	for _, dealID := range dealIDs {
+		deal := st._getOnchainDeal(rt, dealID)
+		st._assertActiveDealState(rt, dealID)
+
+		fee := st._getStorageFeeSinceLastPayment(rt, deal, newPaymentEpoch)
+
+		dealP := deal.Deal().Proposal()
+		st._transferBalance(rt, dealP.Client(), dealP.Provider(), fee)
+
+		// update LastPaymentEpoch in deal
+		deal.Impl().LastPaymentEpoch_ = newPaymentEpoch
+		st.Deals()[dealID] = deal
+	}
+
+	UpdateRelease(rt, h, st)
+}
+
+// unlock remaining payments and return all UnlockedStorageFee to Provider
+// remove deals from ActiveDeals
+// return collaterals to both miner and client
+func (a *StorageMarketActorCode_I) ProcessDealExpiration(rt Runtime, dealIDs []deal.DealID) {
+	h, st := a.State(rt)
+
+	for _, dealID := range dealIDs {
+
+		expiredDeal := st._getOnchainDeal(rt, dealID)
+		st._assertActiveDealState(rt, dealID)
+
+		dealP := expiredDeal.Deal().Proposal()
+		fee := st._getStorageFeeSinceLastPayment(rt, expiredDeal, dealP.EndEpoch())
```

Comment:
should this be all remaining fee? If so, may be worth checking below, lest the _unlockBalance inadvertently leave fees stuck in the deal